### ISSUE TITLE
Subform Support in Royale Conversion

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1192]
On of the big remaining features for Royale conversion is subform support.  See more information about the configuration in #1015

**Comment by JoelProminic** [comment]
I need to provide an example project for this, with examples of nesting.

Note that this is not dependent on a real NSFConverterPortalProject converted database.  We can create the subforms from the mockup editor as well.

**Comment by JoelProminic** [comment]
Here is the example database we discussed in the meeting.  Let me know if you need more guidance for it:
Database:  [TestSubform2.nsf.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/10983288/TestSubform2.nsf.zip)
Converted project:  [TestSubform2_DominoVisualEditor.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/10983290/TestSubform2_DominoVisualEditor.zip)

**Comment by piotrzarzycki21** [comment]
Hey @feather812002 . Can you please explain what is the purpose of that file ? Why it should exists and is needed during conversion ?

**Comment by feather812002** [comment]
@piotrzarzycki21 @JoelProminic 
Okay, if do you mean about this [public static function convertDomino(file:FileLocation):void](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/50b514e3e6163dab1a8a26a1564a5a2d3127dc0f/ide/MoonshineSharedCore/src/actionScripts/plugin/actionscript/as3project/importer/FlashDevelopImporter.as#L335)

The method work for that  project export from NSFConvertProtal , 
The work flow as follow:
1. Customer upload a NSF database and export a Visual project from NSFConvertProtal.
2. Customer open the Moonshine-IDE , then open& import the Visual project with Moonshine-IDE.
    The method in here will convert the Visual project to Domino Form/Subofrm/Page or other some Domino element save as dxl file .
3. Customer just run the build with ODP command with Moonshine-IDE or in the terminal , it should be build back to NSF database.

In short , the method work for auto convert the intermiddle xml that export from NSFConverprotal to dxl file , and let it can build to NSF database again , @JoelProminic know more details about this .
If you have more question about this you can talk with @JoelProminic in the daily meeting direct , he must can explain more good than me .

Thanks.

**Comment by piotrzarzycki21** [comment]
To discuss for todays meeting:
- Where subforms should be placed in `/nsfs/nsf-moonshine/odp`
- Where subforms should be places after conversion to Royale

**Comment by feather812002** [comment]
> * Where subforms should be placed in `/nsfs/nsf-moonshine/odp`

I think for this topic , it only have one answer :  
The subform must should placed into `/nsfs/nsf-moonshine/odp/SharedElements/Subforms`
Then it can success build to NSF database again , if you put the subfrom dxl file in other place, it will break the NSF database build .

**Comment by piotrzarzycki21** [comment]
@JoelProminic as we have discussed I'm attaching example manually modified Royale app.
[TestSubform2_Royale.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/11151000/TestSubform2_Royale.zip)

**Comment by JoelProminic** [comment]
Looking at the project, this is a bit off from the design I was proposing.  I made a modified copy as an example.  My changes:
- The Subform interface extend the interfaces for any nested subforms
- I removed IDominoVO from the subform interfaces.  Subforms will not have separate documents, so they don't have separate DominoUniversalIDs
- Form VOs interfaces should implement the nested subform interfaces.  The Form interface is probably not necessary here
- Remove the Subform references from the Form VOs - the fields will be implemented directly in the VO to satisfy the interface
- Copy the field implementations from the subform VOs to the form VOs
- Remove the Subform VOs - these will be implemented in the Forms VOs
- Update the subform binding in the Form and Subform Views to be the original object rather than a subform reference. Note that the type of itemVO is still does not change - it is the corresponding form or subform interface

The project compiles for me, but there may have been some changes that I missed.  If this needs more explanation, we can discuss it tomorrow:

[TestSubform2_Royale_Modified.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/11152714/TestSubform2_Royale_Modified.zip)

**Comment by piotrzarzycki21** [comment]
> Looking at the project, this is a bit off from the design I was proposing. I made a modified copy as an example. My changes:
> 
> * The Subform interface extend the interfaces for any nested subforms
> * I removed IDominoVO from the subform interfaces.  Subforms will not have separate documents, so they don't have separate DominoUniversalIDs
> * Form VOs interfaces should implement the nested subform interfaces.  The Form interface is probably not necessary here
> * Remove the Subform references from the Form VOs - the fields will be implemented directly in the VO to satisfy the interface
> * Copy the field implementations from the subform VOs to the form VOs
> * Remove the Subform VOs - these will be implemented in the Forms VOs
> * Update the subform binding in the Form and Subform Views to be the original object rather than a subform reference. Note that the type of itemVO is still does not change - it is the corresponding form or subform interface
> 
> The project compiles for me, but there may have been some changes that I missed. If this needs more explanation, we can discuss it tomorrow:
> 
> [TestSubform2_Royale_Modified.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/11152714/TestSubform2_Royale_Modified.zip)

I have pushed to master changes which allow transform sub forms to Royale according to our plan. Let's discuss what's next on our meeting.
